lightning/src/routing/network_graph.rs

```diff
@@ -0,0 +1,1596 @@
+		let features = Readable::read(reader)?;
+		let last_update = Readable::read(reader)?;
+		let rgb = Readable::read(reader)?;
+		let alias = Readable::read(reader)?;
+		let addresses_count: u64 = Readable::read(reader)?;
+		let mut addresses = Vec::with_capacity(cmp::min(addresses_count, MAX_ALLOC_SIZE / 40) as usize);
+		for _ in 0..addresses_count {
+			match Readable::read(reader) {
+				Ok(Ok(addr)) => { addresses.push(addr); },
+				Ok(Err(_)) => return Err(DecodeError::InvalidValue),
+				Err(DecodeError::ShortRead) => return Err(DecodeError::BadLengthDescriptor),
+				_ => unreachable!(),
+			}
+		}
+		let announcement_message = Readable::read(reader)?;
+		Ok(NodeInfo {
+			channels,
+			lowest_inbound_channel_fees,
+			features,
+			last_update,
+			rgb,
+			alias,
+			addresses,
+			announcement_message
+		})
+	}
+}
+
+/// Represents the network as nodes and channels between them
+#[derive(PartialEq)]
+pub struct NetworkGraph {
+	channels: BTreeMap<u64, ChannelInfo>,
+	nodes: BTreeMap<PublicKey, NodeInfo>,
+}
+
+impl Writeable for NetworkGraph {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		(self.channels.len() as u64).write(writer)?;
+		for (ref chan_id, ref chan_info) in self.channels.iter() {
+			(*chan_id).write(writer)?;
+			chan_info.write(writer)?;
+		}
+		(self.nodes.len() as u64).write(writer)?;
+		for (ref node_id, ref node_info) in self.nodes.iter() {
+			node_id.write(writer)?;
+			node_info.write(writer)?;
+		}
+		Ok(())
+	}
+}
+
+impl Readable for NetworkGraph {
+	fn read<R: ::std::io::Read>(reader: &mut R) -> Result<NetworkGraph, DecodeError> {
+		let channels_count: u64 = Readable::read(reader)?;
+		let mut channels = BTreeMap::new();
+		for _ in 0..channels_count {
+			let chan_id: u64 = Readable::read(reader)?;
+			let chan_info = Readable::read(reader)?;
+			channels.insert(chan_id, chan_info);
+		}
+		let nodes_count: u64 = Readable::read(reader)?;
+		let mut nodes = BTreeMap::new();
+		for _ in 0..nodes_count {
+			let node_id = Readable::read(reader)?;
+			let node_info = Readable::read(reader)?;
+			nodes.insert(node_id, node_info);
+		}
+		Ok(NetworkGraph {
+			channels,
+			nodes,
+		})
+	}
+}
+
+impl std::fmt::Display for NetworkGraph {
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+		write!(f, "Network map\n[Channels]\n")?;
+		for (key, val) in self.channels.iter() {
+			write!(f, " {}: {}\n", key, val)?;
+		}
+		write!(f, "[Nodes]\n")?;
+		for (key, val) in self.nodes.iter() {
+			write!(f, " {}: {}\n", log_pubkey!(key), val)?;
+		}
+		Ok(())
+	}
+}
+
+impl NetworkGraph {
+	/// Returns a list of known valid channels
+	pub fn get_channels<'a>(&'a self) -> &'a BTreeMap<u64, ChannelInfo> { &self.channels }
+	/// Returns a list of known nodes
+	pub fn get_nodes<'a>(&'a self) -> &'a BTreeMap<PublicKey, NodeInfo> { &self.nodes }
+
+	fn process_node_announcement(&mut self, msg: &msgs::NodeAnnouncement) -> Result<bool, LightningError> {
```

Comment:
I think these should be pub. Its fine to do it in a separate PR, but they're pretty useful, at least once they have sufficient documentation. May also make sense to not call them "process" as much as add_node_from_announcement or update_node_from_announcement or something like that, making sure to note in the docs that signatures are *not* checked.

Reply:
I will do this, but if we making them public, does it make sense to move should_relay to the message handler, one level above?
It probably doesn't make much sense for every caller to know that, and if they want to relay or not, they can check it on their own, since it's trivial now and won't need db access.